client/web/src/components/externalServices/AddExternalServicesPage.tsx

```diff
@@ -120,102 +124,162 @@ export const AddExternalServicesPage: FC<AddExternalServicesPageProps> = ({
+                </Link>
+                , if you have enabled this feature.
+            </li>
+        </ul>
+        <div className="d-flex justify-content-end">
+            <Button variant="secondary" onClick={dismissPrivacyWarning}>
+                Do not show this again
+            </Button>
+        </div>
+    </Alert>
+)
+
+interface ExternalServicesGroup {
+    services: AddExternalServiceOptionsWithID[]
+    icon: React.ComponentType<{ className?: string }>
+    description: string
+    renderServiceIcon: boolean
+}
+
+type GroupedServiceDisplayName = 'GitHub' | 'GitLab' | 'Bitbucket' | 'Other code hosts'
+
+const computeExternalServicesGroup = (
+    services: Record<string, AddExternalServiceOptions>,
+    allowedCodeHosts: AddExternalServiceOptions[] | null
+): Record<GroupedServiceDisplayName, ExternalServicesGroup> => {
+    const groupedServices: Record<GroupedServiceDisplayName, ExternalServicesGroup> = {
+        GitHub: {
+            services: [],
+            icon: GithubIcon,
+            description: 'Connect with repositories on GitHub',
+            renderServiceIcon: false,
+        },
+        GitLab: {
+            services: [],
+            icon: GitLabIcon,
+            description: 'Connect with repositories on GitLab',
+            renderServiceIcon: false,
+        },
+        Bitbucket: {
+            services: [],
+            icon: BitbucketIcon,
+            description: 'Connect with repositories on Bitbucket',
+            renderServiceIcon: false,
+        },
+        'Other code hosts': { services: [], icon: GitIcon, description: '', renderServiceIcon: true },
+    }
+
+    for (const [serviceID, service] of Object.entries(services)) {
+        const isDisabled = Boolean(allowedCodeHosts && !allowedCodeHosts.includes(service))
+        const otherProps = isDisabled
+            ? {
+                  badge: 'enterprise',
+                  tooltip: 'Upgrade to Sourcegraph Enterprise to add repositories from other code hosts',
+              }
+            : {}
+        switch (service.kind) {
+            case ExternalServiceKind.GITHUB:
+                groupedServices.GitHub.services.push({ ...service, serviceID, enabled: !isDisabled, ...otherProps })
+                break
+            case ExternalServiceKind.GITLAB:
+                groupedServices.GitLab.services.push({ ...service, serviceID, enabled: !isDisabled, ...otherProps })
+                break
+            case ExternalServiceKind.BITBUCKETCLOUD:
+            case ExternalServiceKind.BITBUCKETSERVER:
+                groupedServices.Bitbucket.services.push({ ...service, serviceID, enabled: !isDisabled, ...otherProps })
+                break
+            default:
+                groupedServices['Other code hosts'].services.push({
+                    ...service,
+                    serviceID,
+                    enabled: !isDisabled,
+                    ...otherProps,
+                })
+        }
+    }
```

Comment:
I think this is much shorter. Since the `push` operation is doing the same, no need to repeat it 4 times IMO.

```suggestion
        let key = 'other'
        if (service.kind === ExternalServiceKind.GITHUB || service.kind === ExternalServiceKind.GITLAB) {
            key = service.kind.toLowerCase()
        } else if (service.kind === ExternalServiceKind.BITBUCKETCLOUD || service.kind === ExternalServiceKind.BITBUCKETSERVER) {
            key = 'bitbucket'
        }
        groupedServices[key]services.push({ ...service, serviceID, enabled: !isDisabled, ...otherProps })
    }
```